### ISSUE TITLE
Strava times syncing incorrectly

### ISSUE THREAD

**Issue by The-Maso** [cpfair/tapiriik#117]
Using Runkeeper and syncing to Strava and Endomondo has worked great until recently. Last two runs, times in Endomondo match with Runkeeper but Strava times are way off (about 10% less in total time). The per KM tines all match, but the total has been calculated wrong - and I don't think this is a defect in Strava but a mapping issue.

One run had stops and the other didn't so I can trace it back to that. If you want details I can provide them. or GPX from  files from either I guess.

**Comment by j040p3d20** [comment]
Mine are nearly 100% off... I ran around 45 min on endomondo and on strava it was registered around 25 min. This happened several times.

**Comment by cagnulein** [comment]
Same here: have a look at the original https://www.endomondo.com/workouts/533545727/42759 and to the imported one https://app.strava.com/activities/321748559

**Comment by The-Maso** [comment]
Bump... any update on this one. The issue is still present and quite inconsistent. There's always a difference but sometimes small, sometimes large; always big enough to be unusable syncing in that direction.

Am forced to use Strava app only and sync. Not a big problem, but prefer Runkeeper for day to day use.

**Comment by Drikkie** [comment]
Yeah, I encountered this issue also.
I synced from Endomondo -> Strava, all moving times are way shorter and hence all "pace" is on WR times.

I will check if I ever encountered this before with my cycling entries.

**Comment by niosus** [comment]
@Drikkie, I will briefly duplicate the cause of the issue found in the link you have sent. The idea is that Endomondo generates some duplicate GPS points, Strava, apparently treats them as pauses and so pauses the moving time. This results in a significantly lesser moving time than the actual time.

I have written a small (extremely ugly) proof-of-concept script to remove the duplicates from the exported Endomondo workout. It reads the workout from `input_endo.gpx` and writes the result to `output.gpx`. After the duplicates are removed, one can import the `output.gpx` to Strava and it shows correct time.

The full script:

```
#!/usr/bin/python3
import xml.etree.ElementTree as ET
prevLon = 0.0
prevLat = 0.0
ET.register_namespace('', "http://www.topografix.com/GPX/1/1")
ET.register_namespace('gpxtps', "http://www.garmin.com/xmlschemas/TrackPointExtension/v1")
ET.register_namespace('gpxx', "http://www.garmin.com/xmlschemas/GpxExtensions/v3")
ET.register_namespace('xsi', "http://www.w3.org/2001/XMLSchema-instance")
tree = ET.parse('input_endo.gpx')
root = tree.getroot()
pts_to_remove = []
for trkpt in root[1][3]:
  lon = trkpt.get('lon')
  lat = trkpt.get('lat')
  if (prevLon == lon and prevLat == lat):
    pts_to_remove.append(trkpt)
  else:
    prevLon = lon;
    prevLat = lat;

for pt in pts_to_remove:
  root[1][3].remove(pt)

tree.write('output.gpx')
```

You may see that there is some ugly hardcoding, but I wanted to have it as a proof of concept. I am also not sure where to put it in the current codebase of tapiriik to generate a proper pull request. I will gladly do that if there is a chance that this fix will land into the production version. This is an annoying bug.

**Comment by cagnulein** [comment]
Any followup?

**Comment by niosus** [comment]
@cagnulein, I need to say, that something has changed. I don't exactly know where. I do not experience this issue anymore, as Strava now has all my workouts with correct times imported from Endomondo. Do you experience the bug still? If so I could generate a pull request, but 
a) the tapiriik does not seem to be developed very actively
b) this solution is very hacky (just throwing away data, basically), so I don't know if there is place for it in the production code

**Comment by cagnulein** [comment]
I've just imported this https://www.strava.com/activities/384574868 and the problem is still here :(
If you could, generate a pull request, maybe i could build up a local tapiriik with your mod

**Comment by niosus** [comment]
@cagnulein, ok, sounds good. I will probably look into it on the weekend. It should be a pretty small change, but some time is still needed and I have none during the working week currently :(

**Comment by Drikkie** [comment]
I'm gonna do a new run today and have some saved up.
Gonna sync them today and check how it behaves. I'll let you know.

**Comment by niosus** [comment]
@Drikkie, @cagnulein, that's exactly the point. Something has probably changed in the way endomondo saves the workouts (the only logical explanation as tapiriik did not change, and if we import old workouts they are still corrupted on Strava), because all the new workouts seem to have correct timings now. So do check it out.

**Comment by niosus** [comment]
I have added a pull request. But it is not well tested as the endomondo and strava do not work from tapiriik running locally on my machine (probably some authorization issues). It would be cool if someone could test this. It is probably not the best solution as this will ruin the workouts where you have been standing in one place for some time.

**Comment by Drikkie** [comment]
All new stuff got synced the right way.
Can I try online syncing the old ones, or do I need todo that locally or in a test-version?

**Comment by niosus** [comment]
@Drikkie, I am not the maintainer, so I have just created a pull request. It is not yet (and I don't know if it will ever be) live online. So you would need to do that locally for the old ones. However, if the new ones sync correctly it means that the Endomondo guys have fixed it on their side.

**Comment by Drikkie** [comment]
let's hope it's that. I'm trying to sync new ones. But currently I don't get all of them (though at least one did) resync them after I delete them from Strava. 

I'll keep this updated if old ones that get resynced will also be OK.

**Comment by cagnulein** [comment]
Hi @niosus , i've almost setup my local account but i need an Endomondo Key. So i ask to endomondo and i'm still waiting to receive it.
I'll let you know as soon as endomond will give me access.
Thanks for your work.

**Comment by niosus** [comment]
@cagnulein, great! Waiting for the good news then :+1:

**Comment by cagnulein** [comment]
Endomondo answers me :(

"We are at the moment not performing any new API based integrations. We will keep your information and inform at such a time when we will open up for new requests again."

**Comment by niosus** [comment]
@cagnulein sorry to hear that... :( Then we need someone with endomondo auh token to test this

**Comment by niosus** [comment]
Hey everyone, @cpfair has just merged my pull request, so now it's in master, which probably means that it's gonna be live soon. Which means we won't need the auth token from Endomondo (which they refuse to give).

**Comment by cpfair** [comment]
Yeah - sorry about the delay, busy times in day-job-land :(.

**Comment by niosus** [comment]
@cpfair story of my life :) no problem, keep up the great work! This app is a life saver! :+1:

**Comment by niosus** [comment]
Hmmm, now the average pace is correct: Endomondo: 05:13 min/km, Strava: 5:14 min/km, but the splits are off... :( Strange. Splits are the following.

Endomondo:
![image](https://cloud.githubusercontent.com/assets/1893598/10478694/b0a36574-725f-11e5-94c5-9cf989eb789b.png)

Strava:
![image](https://cloud.githubusercontent.com/assets/1893598/10478711/cf817a62-725f-11e5-81ad-3e9cf77f2c7b.png)

I have a gut feeling like this is related to #179

**Comment by alexmonteiro** [comment]
I've got same issue using Runkeeper and sync with Strava. In my opinion, you can fix it setting the Runkeeper Duration to be the Strava Moving Time. 

Nowadays, the Duration in Runkeeper is set as the the Elapsed Time on Strava, and Moving Time(Strava) are being set with 10% less then Runnkeper duration. 

That's why on Runkeeper I'm doing a 7min/km pace and in Strava shows 5,14km/pace.

Take a look:
Runkeeper running with HR. 
https://runkeeper.com/user/alexmonteiro/activity/675249643

Strava running after sync.
https://www.strava.com/activities/413499408

**Comment by weflyatnight** [comment]
Can you please explain further, have the same error, use Runkeeper and sync to Strava.
All Strava activities are like 10% faster than the actual activity

**Comment by mkornegay2** [comment]
BUMP!!  What is the deal? Two years later and no fix with RK and Strava discrepancies. RK 29:08, Strava 27:48. That's a 1:30 difference and a :30 sec pace difference. Strava is showing PR's all over the place. Really going to screw up my tracking. Not cool.

**Comment by kunickiaj** [comment]
Yep, still experiencing this issue as well

**Comment by enriquebay** [comment]
Same here. Using runkeeper and strava imports mark around 10% faster times.

**Comment by epbx** [comment]
I'd just like to add that I'm having the exact same issues going from Runkeeper to Strava.  When I pointed out the discrepancies to Strava, they told me: " It looks like we encountered some GPS errors in your file that caused us to revert your distance to a GPS-based distance stream instead of device-based" and that "The problem was with the data from your activity and it essentially switched over to a fail safe view for the distance stream".  I reached out to Runkeeper about it and was told: "We read and collect data differently than other apps, and since we don't integrate with Strava–we have no idea how our data will appear over there, sorry about that!  If the issue is how that data is being read, or appearing incorrectly, then this is only something they'd control".  For what it's worth, Strava was able to fix the runs in question so that the data displayed correctly, but it's a bummer that's it's a manual process and neither Runkeeper nor Strava will do anything to resolve it.

**Comment by mkornegay2** [comment]
Yep epbx. Been there and done that. I went to Garmin for my runs and cycles. Apple apps just can't perform consistently and it's ALWAYS someone else issue.
